apps/gatsby/src/Sidebar.js

```diff
@@ -42,8 +40,72 @@ export default class Sidebar extends React.Component {
+    ) : undefined;
+    // If there is no constructor set the url as the base preview
+    if (!constructor){
+      return
+    }
+
+    //QUESTION: What does the locale object in the sdk look like when there are multiple locales available. Need to account for spaces that have more than one locale, so they can have previews everywhere.
+
+    // Recursive helper to return slug values buried in a chain of references
+    const resolveReferenceChain = async (sdk, array, index, parentId) => {
+      console.log(sdk)
+      // Full entry to access child fields
+      const fullParentEntry = await sdk.space.getEntry(parentId)
+      // Child field
+      const childField = fullParentEntry.fields[array[index + 1]][sdk.locales.default]
+      if (Array.isArray(childField)) {
+       console.error("Gatsby Preview App: You are trying to search for a slug in a multi reference field. Only single reference fields are searchable with this app. Either change the field to a single reference, or change the field you are searching for in the slug.")
+        return ""
+      }
+
+      if (index + 2 < array.length) {
+        return resolveReferenceChain(sdk, array, (index + 1), childField.sys.id)
+      } else {
+        return childField
+      }
+    }
+    //Get array of fields to build slug
+    const parentFields = constructor.slug.split("/")
+    //Get array of subfields if the slug is using references
+    const subFields = parentFields.map(parent => parent.split("."))
+    //Generate slug
+    const slug = await (
+      Promise.all(
+        subFields.map(async fieldArray => {
+          try {
+            // If the generated array's length is greater than 1 it means there is reference chain which can be resolved recursively with the resolveReferenceChain function
+            if (fieldArray.length > 1) {
+              const parentId = this.sdk.entry.fields[fieldArray[0]].getValue().sys.id
+              return resolveReferenceChain(this.sdk, fieldArray, 0, parentId)
+            } else if(fieldArray[0].includes('"' || "'")) { // Checks for static text
+              return fieldArray[0].replace(/['"]/g, "")
+            } else { //Field directly on the entry, no use for reference resolver
+              return this.sdk.entry.fields[fieldArray[0]].getValue()
+            }
+          } catch {
+            console.error(`Gatsby Preview App: ${fieldArray[0]}, as defined in the slug field for this content type in the Gatsby Preview App, is not a field. Maybe you mistyped it, or maybe you meant it to be a static string in which case you need to surround it in quotes: ${`"${fieldArray[0]}"`}. The open preview button will send users to your site's base url until fixed.`)
+          }
+        })
+      )
+    )
+    //Make sure the base preview url ends with a /
+    const cleanPreviewUrl = previewUrl.charAt(previewUrl.length - 1) === "/" ? previewUrl : `${previewUrl}/`
+    
+    const fullUrl = `${cleanPreviewUrl}${slug.toString().replace(/,/i, "/")}` 
```

Comment:
Shall we maybe assert the slug has the correct shape and throw if not rather than relying on the regexp-replace output here?

Reply:
Do you mean assert is the TS sense of assert? I think this is all written in JS. Also do you mean throw an error on the frontend, like a notification to the editor?

Reply:
I mean just checking whether the output is conform to what we expect. For example:

```
const assertSlug = (s) => {
  if (slug.length > 0) { 
    return 
  }
  
  throw new Error('Unexpected slug shape')
}
```

And then in the code

```
    const cleanPreviewUrl = previewUrl.charAt(previewUrl.length - 1) === "/" ? previewUrl : `${previewUrl}/`

    assertSlug(slug)

    const fullUrl = `${cleanPreviewUrl}${slug.join("/")}` 
```